lib/factory.rb

```diff
@@ -1,84 +1,101 @@
 # frozen_string_literal: true
 
 class Factory
-  class << self
-    def new(*arguments, &block)
-      name = arguments.shift if arguments.first.is_a? String
+  def self.new(*arguments, &block)
+    name = arguments.shift if arguments.first.is_a? String
 
-      instance = Class.new do
-        attr_accessor *arguments
+    clazz = Class.new do
+      attr_accessor *arguments
 
-        define_method :initialize do |*params|
-          raise ArgumentError, 'Excess arguments' if params.size > arguments.size
+      define_method :initialize do |*params|
+        raise ArgumentError, 'Excess arguments' if params.size > arguments.size
 
-          params.each_with_index do |value, index|
-            instance_variable_set("@#{arguments[index]}", value)
-          end
+        arguments.each_with_index do |_value, index|
+          instance_variable_set("@#{arguments[index]}", params[index])
         end
+      end
 
-        def ==(other)
-          raise TypeError, 'Wrong class' unless self.class == other.class
-          instance_variables_values == other.instance_variables_values
-        end
+      def to_s
+        to_a << self.class
+      end
 
-        def [](param)
-          var = if param.is_a? Integer
-                  instance_variables[param]
-                else
-                  "@#{param}"
-                end
-          instance_variable_get(var)
-        end
+      def ==(other)
+        to_s == other.to_s
+      end
 
-        def []=(param, value)
-          instance_variable_set("@#{param}", value)
+      def [](param)
+        attr_name = if param.is_a?(Integer)
+                      raise IndexError unless instance_variables[param]
+                      instance_variables[param]
+                    elsif param.is_a?(Float)
+                      raise IndexError unless instance_variables[param.floor]
+                      instance_variables[param.floor]
+                    else
+                      "@#{param}"
         end
+        raise NameError unless instance_variable_get(attr_name)
```

Comment:
If you have an instance variable and its value is `nil` or `false`, then `NameError` will be raised. It's incorrect behaviour. Do you know why?